### ISSUE TITLE
Tweak reference counting?

### ISSUE THREAD

**Issue by lydell** [elm-tooling/elm-language-server#181]
## Expected Behavior
Above functions, "4 references" or similar can be shown. "4" should mean that there are 4 places where the function is being used in "actual" code.

## Current Behavior
"4" references means that there are 4 occurances of the function name (that actually resolve to this function).

## Steps to Reproduce (for bugs)

Starting code:

```elm
module Main exposing (..)

import Math


result =
    Math.add 1 2

```

```elm
module Math exposing (add)


add : number -> number -> number
add =
    (+)
```

Above the `add` function in Math.elm, this text is shown:

> exposed | 4 references

But in my opinion it only has 1 reference – the call in Main.elm.

In other words, these places don’t count:

- The definition of the function.
- The type annotation of the function.
- `exposing`, both in `module` and `import` lines

There’s one edge case I can think of: If a file has `import Math exposing (add)` but never uses `add` – does that count?

## Context
I think the `exposed | X references` feature is quite nice. But a function that is never used should get `0 references`, shouldn’t it?

## Your Environment

- Version used: 0.6.0
- Editor name and version: VSCode 1.39.2
- Environment name and version: Node.js v12.8.0
- Operating System and version: macOS Mojave

**Comment by Augustin82** [comment]
Agree it "should" be like OP says, but it's not critical either way.
Note: I stopped using that feature pretty soon after my adoption of ElmLS, so take my opinion with a grain of salt.

**Comment by manuscrypt** [comment]
I agree. and if it isn't used from anywhere, the function could be made slightly less visible, as e.g. resharper does in c#/vsstudio

**Comment by Razzeee** [comment]
That's planned, but depends on the next version of the protocol we use. So
we need to wait for that.

**Comment by Razzeee** [comment]
I pushed https://github.com/elm-tooling/elm-language-server/tree/rework-reference-codelens which changes the counting. But it's still kinda weird.

Count shows 1. You click the reference list and get to peek references, which is only one.

If we click a reference and say "Find all references" we find all references, including the ones from type annotations, exposing/importing and definitions. 

So which one is wrong? Should we change "Find all references"?

**Comment by manuscrypt** [comment]
Maybe differentiate between "Find all usages" and "Find all references"?

**Comment by lydell** [comment]
When I use "Find all references" it is to answer the question "Who is using this?"

**Comment by manuscrypt** [comment]
I hear you and counter by throwing in the term "textual occurrances" without further elaboration.

**Comment by lydell** [comment]
Yes, that’s another valid definition. But is it helpful?

**Comment by manuscrypt** [comment]
No, I still agree with your viewpoint on how the feature should behave. But if you ask "who is using this" then one could argue that the module uses it to make the relationship work. Just playing devil's advocate here.

**Comment by Razzeee** [comment]
So in typescript a line like
```
import { DocumentSymbolProvider } from "./providers/documentSymbolProvider";
```
shows up as a reference for `DocumentSymbolProvider`. I don't think we have constructs similar to type annotations and module expose in TS.

**Comment by Razzeee** [comment]
I'm not sure how to progress on this. Maybe no people leaving comments on this the past months is enough to verify the current handling as okay?

**Comment by jmbockhorst** [comment]
I think not including the type annotation in the references makes sense, as it doesn't really add anything useful.

**Comment by manuscrypt** [comment]
So we have three different behaviors, if I summarize correctly:
1. Count every occurrence of the symbol name, as in "rename symbol"
2. Count only places, where the symbol name is referenced, which could be:
   2a) including exports, imports
   2b) including function bodies (thinking also about ports here)
3. Count only places, where changing the type definition / function signature would have an impact (the term "usages" might apply here). This might be useful for automatic refactoring (remove parameters, or add TODO placeholders)

2b could be seen as a subset of 3, therefore I lean towards 3.

**Comment by Razzeee** [comment]
> I think not including the type annotation in the references makes sense, as it doesn't really add anything useful.

I feel like it does (which is the reason I choose that design and not just to be able to use that for rename code)
I'm mostly split on including in in the count, which then would make the reference list inconsistent to the count, which might be very confusing.

**Comment by Razzeee** [comment]
I don't think investing more time into this will do any good. At this point most people seem to have accepted how it's counted, so I'll close this.
